src/dataconnect/schemaMigration.ts

```diff
@@ -55,29 +56,35 @@ export async function migrateSchema(args: {
   if (!instanceId) {
     throw new FirebaseError(`tried to migrate schema but ${instanceId} was undefined`);
   }
+  const serviceName = schema.name.replace(`/schemas/${SCHEMA_ID}`, "");
   try {
-    const serviceName = schema.name.replace(`/schemas/${SCHEMA_ID}`, "");
     await ensureServiceIsConnectedToCloudSql(serviceName);
     await upsertSchema(schema, validateOnly);
     logger.debug(`Database schema was up to date for ${instanceId}:${databaseId}`);
-    return [];
   } catch (err: any) {
-    const incompatible = getIncompatibleSchemaError(err);
-    if (!incompatible) {
+    const incompatible = errors.getIncompatibleSchemaError(err);
+    const invalid = errors.isInvalidConnectorError(err);
+    if (!incompatible && !invalid) {
       // If we got a different type of error, throw it
       throw err;
     }
-    // Try to migrate schema
-    const diffs = await handleIncompatibleSchemaError({
-      ...args,
-      incompatibleSchemaError: incompatible,
-      instanceId,
-      databaseId,
-    });
-    // Then, try to upsert schema again. If there still is an error, just throw it now
-    await upsertSchema(schema, validateOnly);
-    return diffs;
+    if (invalid) {
```

Comment:
I think the returning both invalid connectors and schema migration errors in the backend is at risk to make I/O.

We can do a stop gap solution to call `migrateSchema` twice in all callers, so the second call will capture the "invalid connector" errors. Still perform schema migration before recreate connectors, but this adds a prompt between (a) and (b) above. Acceptable stop gap while we improve backend.

Reply:
First, to clarify, this does not delete connectors during `sql:migrate` as is (since validateOnly=true). It would be reasonable to cut off before offering any database schema migration during 'sql:migrate' if there are invalid , I can make that change.

Good call on splitting the confirmations and the actual changes - will do so

Reply:
Made a pretty large refactor here to match this new beahvior, PTAL. I tested a bunch of cases locally and the behavior now matches what you've described here.

Reply:
Thanks a bunch for the refactor! Looks great!